test/framework/clusterctl/clusterctl_helpers.go

```diff
@@ -274,42 +282,67 @@ func ApplyClusterTemplateAndWait(ctx context.Context, input ApplyClusterTemplate
 	Expect(ctx).NotTo(BeNil(), "ctx is required for ApplyClusterTemplateAndWait")
 	Expect(input.ClusterProxy).ToNot(BeNil(), "Invalid argument. input.ClusterProxy can't be nil when calling ApplyClusterTemplateAndWait")
 	Expect(result).ToNot(BeNil(), "Invalid argument. result can't be nil when calling ApplyClusterTemplateAndWait")
-	Expect(input.ConfigCluster.ControlPlaneMachineCount).ToNot(BeNil())
-	Expect(input.ConfigCluster.WorkerMachineCount).ToNot(BeNil())
+	if input.WorkloadClusterTemplate == nil {
+		Expect(input.ConfigCluster.ControlPlaneMachineCount).ToNot(BeNil())
+		Expect(input.ConfigCluster.WorkerMachineCount).ToNot(BeNil())
+	}
+	if input.ConfigCluster.ControlPlaneMachineCount == nil {
+		Expect(input.WorkloadClusterTemplate).NotTo(BeNil())
+	}
 
-	log.Logf("Creating the workload cluster with name %q using the %q template (Kubernetes %s, %d control-plane machines, %d worker machines)",
-		input.ConfigCluster.ClusterName, valueOrDefault(input.ConfigCluster.Flavor), input.ConfigCluster.KubernetesVersion, *input.ConfigCluster.ControlPlaneMachineCount, *input.ConfigCluster.WorkerMachineCount)
+	if input.WorkloadClusterTemplate == nil {
+		log.Logf("Creating the workload cluster with name %q using the %q template (Kubernetes %s, %d control-plane machines, %d worker machines)",
+			input.ConfigCluster.ClusterName, valueOrDefault(input.ConfigCluster.Flavor), input.ConfigCluster.KubernetesVersion, *input.ConfigCluster.ControlPlaneMachineCount, *input.ConfigCluster.WorkerMachineCount)
+	} else {
+		log.Logf("Creating the worklog cluster with name %q form the provided yaml", input.WorkloadClusterTemplate.ClusterName)
+	}
+
+	var clusterName, namespace string
```

Comment:
Let's move this up above  l.293 and then use something like "of/for cluster %s" ... klog.KRef in all logs consistently

(logs in 345++ look good to me)